Added page about SNS and Messenger (#16)
Happyr/developer.happyr.com: _posts/2019-06-24-symfony-messenger-on-aws-lambda.md

---
title: Symfony Messenger on AWS Lambda
author: Tobias Nyholm
date: '2019-03-24 09:13:00 +0200'
header:
  image: "images/posts/non-blocking-fonts.png"
  teaser: images/posts/non-blocking-fonts.png
categories:
- Bref
- Messenger
- Symfony
---

For the past 4 years we have developed web applications with a message bus. The Symfony Messenger component was released
around a year ago and I've loved it since. In the past few weeks we've been deploying applications on AWS Lambda. Going
serverless is super cool since you can scale up and down, it is cheap bla bla bla. The major selling point to me is that 

> you dont need to care about dev ops once it all up and running. We've been using [Bref](https://bref.sh) and I can highly 

Reviewer: ```suggestion
you don't need to care about dev ops once it all up and running. We've been using [Bref](https://bref.sh) and I can highly 
```

recommend it. 

> When running your application on Lambda, most things works just as normal but there are one thing in particular to be aware

Reviewer: ```suggestion
When running your application on Lambda, most things work just as normal but there is one thing in particular to be aware
```

of. Lambda is not meant to be running and pull on a queue. But that is exactly what is expected when using Symfony's Messenger

> component with async messages. You are expected to run `bin/console messenger:consome` and keep that alive with Supervisord. 

Reviewer: ```suggestion
component with async messages. You are expected to run `bin/console messenger:consume` and keep that alive with Supervisord. 
```

Instead, the queue should make sure to start a Lambda function. 

So on AWS there are two major queue like services: SQS and SNS. They are pretty similar but SQS is more of a classic queue
like RabbitMQ. SNS is more of a publish-subscribe system. We've decided to use SNS because pub-sub is pretty much what we 

> want to do. We dont have the issue that "the server might get too busy" which you would have if you ran on a normal server. 

Reviewer: ```suggestion
want to do. We don't have the issue that "the server might get too busy" which you would have if you ran on a normal server. 
```

        buses:
            # ...
        routing:
            App\Message\Event\UserCreated: sns

enqueue:
    foobar:
        transport:
            dsn: "sns://foo" # Must start with "sns:". Topic details come from framework.mesenger.transport.sns.dsn
            connection_factory_class: 'Enqueue\Sns\SnsConnectionFactory'
            key: '%env(AWS_KEY)%'
            secret: '%env(AWS_SECRET)%'
            region: '%env(AWS_TARGET_REGION)%'
        client: ~
{% endhighlight %}

This will successfully publish messages to the SNS topic named "my_sns_topic". Now we need to create a new Lambda function
to consume the messages. We will use the exact same source code but deployed with this Sams template. 

{% highlight yaml %}
Resources:
    Consumer:
        Type: AWS::Serverless::Function
        Properties:
            FunctionName: 'my-app-consumer'
            Handler: bin/message-consumer
            Timeout: 20 # in seconds
            MemorySize: 2048
            # ...
            Events:
                Sns:
                    Type: SNS
                    Properties:
                        Topic: arn:aws:sns:eu-central-1:123456789:my_sns_topic
{% endhighlight %}

> This will invoke a a script att `bin/consumer` for every message pushed to "my_sns_topic". The job of this script is to 

Reviewer: ```suggestion
This will invoke a script at `bin/consumer` for every message pushed to "my_sns_topic". The job of this script is to 
```

read the SNS event and give it back to the SNS transformer. 

> Let start with defining the `bin/consume` script:

Reviewer: ```suggestion
Lets start by defining the `bin/consume` script:
```

{% highlight php %}
#!/usr/bin/env php
<?php

use App\Kernel;
use App\Consumer\SnsConsumer;

require dirname(__DIR__).'/config/bootstrap.php';
require dirname(__DIR__).'/vendor/autoload.php';

lambda(static function (array $event) {

    $kernel = new Kernel($_SERVER['APP_ENV'], (bool) $_SERVER['APP_DEBUG']);
    $kernel->boot();
    $container = $kernel->getContainer();

    /** @var SnsConsumer $consumer */
    $consumer = $container->get(SnsConsumer::class);

    foreach ($event['Records'] as $record) {

>         if (!isset($record['Sns']) && !isset($record['Sns']['Message'])) {

Reviewer: Can be shortened to: `if (!isset($record['Sns']['Message'])) {`
